src/shared/Microsoft.Git.CredentialManager/Authentication/MicrosoftAuthentication.cs

```diff
@@ -111,41 +121,86 @@ private async Task<JsonWebToken> GetAccessTokenInProcAsync(string authority, str
             //
             // The device code flow has no limitations other than a way to communicate to the user the code required to authenticate.
             //
-
-            // If the user has disabled interaction all we can do is fail at this point
-            ThrowIfUserInteractionDisabled();
-
             if (result is null)
             {
-#if NETFRAMEWORK
-                // If we're in an interactive session and on .NET Framework, let MSAL show the WinForms-based embeded UI
-                if (Context.SessionManager.IsDesktopSession)
+                // If the user has disabled interaction all we can do is fail at this point
+                ThrowIfUserInteractionDisabled();
+
+                // Check for a user flow preference
+                MicrosoftAuthenticationFlowType flowType = GetFlowType();
+                switch (flowType)
                 {
-                    result = await app.AcquireTokenInteractive(scopes)
-                        .WithPrompt(Prompt.SelectAccount)
-                        .WithUseEmbeddedWebView(true)
-                        .ExecuteAsync();
+                    case MicrosoftAuthenticationFlowType.Auto:
+                        if (CanUseEmbeddedWebView())
+                            goto case MicrosoftAuthenticationFlowType.EmbeddedWebView;
+
+                        if (CanUseSystemWebView(app, redirectUri))
+                            goto case MicrosoftAuthenticationFlowType.SystemWebView;
+
+                        // Fall back to device code flow
+                        goto case MicrosoftAuthenticationFlowType.DeviceCode;
```

Comment:
The existence of `goto`s makes me want to restructure this code. However, this `switch` is so carefully constructed, that the alternatives actually look worse. Kudos on finding a good use for `goto`.

Reply:
Yeah, this sort of "have several options, and an 'auto' choice that may pick one of those options" pattern is best served with a `switch` with `goto case`, IMO.
It's the only time I've found `goto` is acceptable; plus `goto case` feels safer that a random label.

This same pattern is already present in the `GitHubAuthentication` class for selecting the type of auth flow (user/pass or OAuth) or letting it auto-detect.